design-system/src/theme-provider.tsx

```diff
@@ -47,38 +31,46 @@ const changeTheme = ({
   themeOverrides,
 }: TChangeTheme): void => {
   const target = isBrowser ? parentSelector() : null;
-  const validTheme = validateTheme(newTheme);
-  target?.setAttribute('data-theme', validTheme);
+
+  // With no target we can't change themes
+  if (!target) return;
+
+  const validTheme = (
+    allThemesNames.includes(newTheme || '') ? newTheme! : 'default'
+  ) as ThemeName;
+  if (newTheme !== validTheme) {
+    console.warn(
+      `ThemeProvider: the specified theme '${newTheme}' is not supported.`
+    );
+  }
+
   const vars = toVars(
     themeOverrides && isObject(themeOverrides)
       ? merge(themes[validTheme], themeOverrides)
       : themes[validTheme]
   );
 
   Object.entries(vars).forEach(([key, value]) => {
-    target?.style.setProperty(key, value);
+    target.style.setProperty(key, value);
   });
+  target.dataset.theme = validTheme;
 };
 
 type ThemeProviderProps = {
+  parentSelector: typeof defaultParentSelector;
   theme?: string;
   themeOverrides?: Record<string, string>;
-  parentSelector: typeof defaultParentSelector;
 };
 
 const ThemeProvider = (props: ThemeProviderProps) => {
-  const parentSelector = useCallback(
-    () => props.parentSelector(),
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    []
-  );
+  const parentSelectorRef = useRef(props.parentSelector);
   useLayoutEffect(() => {
     changeTheme({
       newTheme: props.theme,
-      parentSelector,
+      parentSelector: parentSelectorRef.current,
       themeOverrides: props.themeOverrides,
     });
-  }, [props.theme, props.themeOverrides, parentSelector]);
+  }, [props.theme, props.themeOverrides]);
```

Comment:
Thanks, this is a great improvement 🙌 

As discussed, perhaps we would need to also use [the latest ref pattern](https://epicreact.dev/the-latest-ref-pattern-in-react/) for certain scenarios (one I can think of is having a select for choosing a theme and a separate button for confirming the choice). I'm just leaving this for reference if we need to get back to this in the future.

I guess when we update to React v18 and [`useEvent`](https://github.com/facebook/react/pull/24505) hook is released, that would be a candidate for refactor.

Reply:
I agree this is something we might check later when we support theming change.

On the other hand, I'm sorry I don't think I understand the scenario you mention.
I think the problem with no having the pattern you mention is expecting the users to change the value of `parentSelector` provided to the hook in different renders of their components.

Maybe we can discuss about it in a quick meeting if you want to.

Reply:
Yes, you're right about my scenario